### ISSUE TITLE
False positives for StopTimeTimepointWithoutTimesNotice

### ISSUE THREAD

**Issue by mcplanner-zz** [MobilityData/gtfs-validator#954]
## **Bug report** 

**Describe the bug**
The StopTimeTimepointWithoutTimesNotice error is displayed when the stop_times.timepoint field is blank. This is unexpected. I would expect this to be interpreted as having no timepoints.

**How we reproduce the bug**
Steps to reproduce the behaviour:
This happened with this field: http://iportal.sacrt.com/GTFS/Unitrans/google_transit.zip

**Expected behaviour**
If the timepoint field isn't being used, I expect an error about it's being missing.

**Observed behaviour**
I was alerted that every timepoint was lacking an arrival and departure time.

**Screenshots:** 

**Environment versions**
Java 11, v2.0.0 validator, https://github.com/cal-itp/gtfs-validator

**Comment by lionel-nj** [comment]
Thanks @mcplanner for opening.

> The StopTimeTimepointWithoutTimesNotice error is displayed when the stop_times.timepoint field is blank. This is unexpected. I would expect this to be interpreted as having no timepoints.

The GTFS specification says that if `stop_times.timepoint = 1 or empty` then the times are considered exact. From my understanding a stop time with no value for `timepoint` and no value for `stop_times.arrival_time` or `stop_times.departure_time` should issue a `StopTimeTimepointWithoutTimesNotice`.

See [stop_times.txt reference](https://github.com/google/transit/blob/master/gtfs/spec/en/reference.md#stop_timestxt).

> If the timepoint field isn't being used, I expect an error about it's being missing.

I am not sure if a notice should be generated here, the specification handles the case where no value is attributed to `stop_times.timepoint`: default value is 1.

**Comment by barbeau** [comment]
Here's example data from the above GTFS file:

| trip_id | arrival_time | departure_time | stop_id | stop_sequence | stop_headsign | pickup_type | drop\_off\_type | shape\_dist\_traveled |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 784678 | 07:25:00 | 07:25:00 | 22256 | 1   |     |     |     |     |
| 784678 |     |     | 22239 | 2   |     |     |     |     |
| 784678 | 07:28:00 | 07:28:00 | 22361 | 3   |     |     |     |     |
| 784678 |     |     | 22225 | 4   |     |     |     |     |

tl;dr - This data is fine and the validator shouldn't be logging any errors. The reason is that this data conforms to the original GTFS spec prior to the `timepoint` field existing (`timepoint` is optional). The original GTFS spec said that you should only provide times for records that were timepoints. So in the above data stop_sequence 1 and 3 are timepoints, and 2 and 4 are not.

IMHO the timepoint definition really needs to be clarified in the spec as it currently doesn't communicate this concept clearly - see https://github.com/google/transit/issues/61. 

@MobilityData/transit-specs

**Comment by barbeau** [comment]
>If the timepoint field isn't being used, I expect an error about it's being missing.

The GTFS Best Practices (http://gtfs.org/best-practices/#stop_timestxt) say:

>The timepoint field should be provided. It specifies which stop_times the operator will attempt to strictly adhere to (timepoint=1), and that other stop times are estimates (timepoint=0).

Sticking to the canonical validator definition, timepoint field missing couldn't be an error, but could certainly be a warning.

**Comment by lionel-nj** [comment]
Thanks for the clarification @barbeau! Following [your suggestion](https://github.com/google/transit/issues/61#issuecomment-304999391), I agree that the spec is not clear on this particular point. I'll keep this issue open so that we can fix this bug.

Does the following seems reasonable?

1. Identify if `stop_times.txt` is a legacy dataset: does it contain column `timepoint` or not?  This means that we'll have to modify the `core` module to include `hasTimepointColumn` (or similar) boolean to `GtfsStopTimeTableContainer`. 
2. Remove defaut value `1` from `GtfsStopTimeSchema` for field `timepoint`: 
https://github.com/MobilityData/gtfs-validator/blob/464ff36316162e71d7f154e64e00d87b21f91dd4/main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsStopTimeSchema.java#L77
3. Issue a notice (warning) if a row from `stop_times.txt` has value `timepoint=1`  and no value for `stop_times.arrival_time` and/or `stop_times.departure_time`.

**Comment by lionel-nj** [comment]
> Here are the cases to cover.
> 
> For all stop_times.txt files:
> * First and last record for each trip must have arrival and departure time, else ERROR

✅ Covered in `MissingTripEdgeValidator`.

> * If arrival time or departure time is populated for a record, then both arrival time AND departure time must be populated for that record, else ERROR

✅ Covered in `StopTimeArrivalAndDepartureTimeValidator`. 

> If timepoint column does not exist:
> * Emit WARNING that timepoint column should be added
> 
> If timepoint column exists:
> * All records must have 0 or 1 value in timepoint column, else ERROR
> * If timepoint==1, arrival and departure time must both be populated, else ERROR

> There is also the case where a stop_times.txt file does not have the timepoint column, but all records have arrival and departure times populated. Technically this was against the original GTFS spec, but it became a common practice, and therefore we shouldn't consider it an error because it will invalidate a lot of old datasets. In this case we'll emit a warning anyway that timepoint should be added.

⏳ `TimepointTimeValidator` should be modified to this extent. 

> Does that make sense?

Yes, @isabelle-dr should the specification be clarified first or could these changes be implemented before?

**Comment by isabelle-dr** [comment]
Hey all, thanks for flagging and looking into this. 

The case of legacy data with no timepoint column (prior to the timepoint field existing) was [flagged](https://github.com/MobilityData/gtfs-validator/pull/798#discussion_r592540297) when we initially introduced that rule.

I agree that the root problem is that **the specification contains two contradictory statements about the meaning of an "empty" value in `stop_times.timepoint`**
1. in `stop_times.timepoint`:

> 1 or empty - Times are considered exact.

2. in `stop_times.departure_time` and `stop_times.arrival_time`:

> Conditionally Required:
> Required for timepoint=1.
> Optional otherwise.

The former could mean that`stop_times.timepoint` == 1 is equivalent than `stop_times.timepoint` == "".
The latter could mean that the times are required when `stop_times.timepoint` = 1 and are not required when `stop_times.timepoint` = "". 
If we consider that an empty timepoint column is equivalent to empty values in all `stops_times.timepoint` fields, all legacy datasets become invalid, which is how it was coded in the validator and what @mcplanner-zz is flagging here. She also points out how counter-intuitive it is that having no values for timepoint is considered a timepoint.

This is extremely well described in [google/transit/issues/61](https://github.com/google/transit/issues/61).
The "empty" word in the first statement seems to actually refer to an empty timepoint **column** which maintains the backwards compatibility.

**How should be the validator behaviour?**

> If timepoint column does not exist:
> 
> Emit WARNING that timepoint column should be added

Absolutely, and this will be a new notice based on the Best Practices, that we should add at some point. 

> If timepoint column exists:
> 
> All records must have 0 or 1 value in timepoint column, else ERROR
> If timepoint==1, arrival and departure time must both be populated, else ERROR

I would only implement the second statement proposed here (at least for now) because the first is stricter than the specification.

**Comment by isabelle-dr** [comment]
Here are some examples of what would change, @barbeau does this seem right?

#### Case 1: legacy dataset
**Current behaviour:** The notice is emitted for stops 2 & 3
**Expected behaviour:** No notice
stop_sequence | arrival_time | departure_time
-- | -- | --
1 | 00:00:00 | 00:00:00
2 |   |  
3 |   |  
4 | 00:10:00 | 00:10:00

#### Case 2: dataset with timepoint column, filled
**Current behaviour:** No notice
**Expected behaviour:** No notice
stop_sequence | arrival_time | departure_time | timepoint
-- | -- | -- | --
1 | 00:00:00 | 00:00:00 | 1
2 | 00:02:00 | 00:02:00 | 0
3 | 00:08:00 | 00:08:00 | 0
4 | 00:10:00 | 00:10:00 | 1

#### Case 3: invalid data
**Current behaviour:** The StopTimeTimepointWithoutTimesNotice is emitted for stop 2
**Expected behaviour:** The StopTimeTimepointWithoutTimesNotice is emitted for stop 2
stop_sequence | arrival_time | departure_time | timepoint
-- | -- | -- | --
1 | 00:00:00 | 00:00:00 | 1
2 |  |  | 1
3 | 00:08:00 | 00:08:00 | 0
4 | 00:10:00 | 00:10:00 | 1

#### Case 4 (edge case): dataset with timepoint columns, only 1's, all times populated
**Current behaviour:** No notice
**Expected behaviour:** No notice
stop_sequence | arrival_time | departure_time | timepoint
-- | -- | -- | --
1 | 00:00:00 | 00:00:00 | 1
2 | 00:02:00 | 00:02:00 | 
3 | 00:08:00 | 00:08:00 | 
4 | 00:10:00 | 00:10:00 | 1

#### Case 5 (edge case): dataset with timepoint column, one 1's and some empty times
**Current behaviour:** The StopTimeTimepointWithoutTimesNotice is emitted for stops 2  and 3
**Expected behaviour:** No notice
stop_sequence | arrival_time | departure_time | timepoint
-- | -- | -- | --
1 | 00:00:00 | 00:00:00 | 1
2 |  |  | 
3 |  |  | 
4 | 00:10:00 | 00:10:00 | 1

**Comment by barbeau** [comment]
@isabelle-dr Cases 1, 2, and 3 LGTM 👍 

>I would only implement the second statement proposed here (at least for now) because the first is stricter than the specification: it would make the following data invalid, and there is nothing in spec currently that would justify that.

So Cases 4 and 5 fall into this category. I agree that the current language in the spec is fuzzy around this, although the original intent when introducing the `timepoint` column was to specify all or no values - obviously in hindsight it would have been better if it said that explicitly.

Do we know if there are real datasets in the wild that are partially populating `timepoint` like Case 4 and 5? If so, then I agree that the spec probably needs to be clarified before classifying this as an error. If not, IMHO it could go either way, and IMHO it should be an error.
